### PR TITLE
SAMBRO: admin subscriptions

### PR DESCRIPTION
**Pull request by biplovbhandari**
@flavour Please merge

### CODE REVIEW

File: modules/s3db/pr.py [hunk 1]
```diff
@@ -4072,6 +4080,29 @@ def model(self):
...
+            label_create = T("Create Subscription"),
+            title_display = T("Subscription Details"),
+            title_list = T("Person Subscriptions"),
+            title_update = T("Edit Subscription"),
+            label_list_button = T("List Person Subscriptions"),
+            label_delete_button = T("Delete Person Subscription"),
+            msg_record_created = T("Subscription added"),
+            msg_record_modified = T("Subscription updated"),
+            msg_record_deleted = T("Subscription removed"),
+            msg_list_empty = T("No Subscription currently registered")
```
Comment (flavour) [inline modules/s3db/pr.py]: Subscriptions ;)

File: modules/templates/SAMBRO/config.py [hunk 2]
```diff
@@ -271,6 +271,41 @@ def custom_prep(r):
     settings.customise_sync_repository_controller = customise_sync_repository_controller
 
     # -------------------------------------------------------------------------
+    def customise_pr_subscription_controller(**attr):
+
+        from gluon.html import URL, A
+        from s3 import S3CRUD
+        s3 = current.response.s3
+
+        s3.filter = (current.s3db.pr_subscription.owned_by_group != None)
```
Comment (flavour) [inline modules/templates/SAMBRO/config.py]: Useful to add a comment explaining this (whilst we still remember!)

File: modules/templates/SAMBRO/controllers.py [hunk 3]
```diff
@@ -626,6 +977,89 @@ def _get_subscription(self):
...
+                            continue
+                        if k in get_vars:
+                            if type(get_vars[k]) is list:
+                                get_vars[k].append(v)
+                            else:
+                                get_vars[k] = [get_vars[k], v]
+                        else:
+                            get_vars[k] = v
+
+        output.update({"get_vars": get_vars,
```
Comment (flavour) [inline modules/templates/SAMBRO/controllers.py]: Why not just define output here?
Seems very odd to define empty at the start & then update here?
(This pattern is useful where there are several updates, some of which at least are optional)

File: modules/templates/SAMBRO/menus.py [hunk 4]
```diff
@@ -69,14 +70,14 @@ def menu_modules(cls):
...
                     menus_.extend([homepage("cap"),
                                   homepage("gis"),
                                   ])
-                elif auth.s3_has_role("ALERT_EDITOR") or \
-                     auth.s3_has_role("ALERT_APPROVER"):
-                    menus_.append(homepage("cap"),
-                                  )
+                elif has_role("ALERT_EDITOR") or \
+                     has_role("ALERT_APPROVER"):
+                    menus_.extend([homepage("cap"),
```
Comment (flavour) [inline modules/templates/SAMBRO/menus.py]: Why did you change this to extend here?
append is likely to be more efficient when adding just 1 entry

File: modules/templates/SAMBRO/menus.py [hunk 5]
```diff
@@ -103,11 +104,19 @@ def menu_auth(cls, **attr):
         else:
             # Logged-in
             user_id = auth.s3_logged_in_person()
+            
+            if not (auth.s3_has_role("ALERT_EDITOR") or \
+                    auth.s3_has_role("ALERT_APPROVER") or \
+                    auth.s3_has_role("ADMIN")):
```
Comment (flavour) [inline modules/templates/SAMBRO/menus.py]: You're not using the has_role shortcut here yet

File: modules/templates/SAMBRO/menus.py [hunk 6]
```diff
@@ -165,6 +174,16 @@ def cap():
                                 s3_has_role("ALERT_APPROVER")
 
         return M(c="cap")(
+                    M("Manage Recipient",
```
Comment (flavour) [inline modules/templates/SAMBRO/menus.py]: Recipients?

**Comment by flavour** [comment]
A few minor tweaks& then this can go in finally ;)

**Comment by biplovbhandari** [comment]
@flavour Done :)

File: modules/templates/SAMBRO/config.py [hunk 7]
```diff
@@ -271,6 +271,42 @@ def custom_prep(r):
     settings.customise_sync_repository_controller = customise_sync_repository_controller
 
     # -------------------------------------------------------------------------
+    def customise_pr_subscription_controller(**attr):
+
+        from gluon.html import URL, A
+        from s3 import S3CRUD
+        s3 = current.response.s3
+
+        # Filter admin based subscription
```
Comment (flavour) [inline modules/templates/SAMBRO/config.py]: Post-merge: "Filter to just Admin-controlled subscriptions"
